### DISCUSSION TITLE
[JOBS] AMQP redialer does not resume pipeline (sometimes?)

### DISCUSSION THREAD

**Discussion by algirdasci** [roadrunner-server/roadrunner#2012] [Q&A] [answered]
Hello,

I'm using RR (via Spiral framework) mainly for RabbitMQ message processing. Problem is, that network connection is not very stable, as RR is running in multiple (100+) servers across the globe. I've configured high timeout values, so RR would eventually reconnect to RabbitMQ server and continue to process messages without intervention. Currently each server consumes at least 4 different queues. The pipelines for those queues are created dynamically:

```PHP
    public function declareQueue(string $queueName, array $options = []): void
    {
        $jobs = new Jobs(RPC::create(RPCConst::CONNECTION));
        if (!$this->doesQueueExist($jobs, $queueName)) {
            $jobs->create(self::getAMPQCreateInfo($queueName, $options));
        }
    }

    /**
     * @throws JobsException
     */
    private function doesQueueExist(Jobs $jobs, string $queueName): bool
    {
        foreach ($jobs->getIterator() as $queue) {
            if ($queue->getName() === $queueName) {
                return true;
            }
        }

        return false;
    }

    public static function getAMPQCreateInfo(string $queueName, array $options = []): AMQPCreateInfo
    {
        $nodeQueueName = sprintf('%d.%s', env('SERVER_ID'), $queueName);

        return new AMQPCreateInfo(
            name: $queueName, // @phpstan-ignore argument.type
            priority: $options['priority'] ?? CreateInfo::PRIORITY_DEFAULT_VALUE,
            prefetch: $options['prefetch'] ?? AMQPCreateInfo::PREFETCH_DEFAULT_VALUE,
            queue: $options['queue_name'] ?? $nodeQueueName,
            exchange: QueueConst::EXCHANGE_NAME,
            exchangeType: ExchangeType::Topics,
            routingKey: $options['queue_name'] ?? $nodeQueueName,
            durable: $options['durable'] ?? true,
            exchangeDurable: true,
            consumeAll: $options['consume_all'] ?? false,
            queueHeaders: array_merge($options['queue_headers'] ?? [], [
                'x-queue-type' => 'quorum'
            ]),
            redialTimeout: 2 * 24 * 3600
        );
    }
```

Everything is working ok, but there are some cases, when redialer restores connection to RabbitMQ and leaves the pipeline paused, which means that message processing stops and RR needs to be restarted or job queue resumed, to continue working. Also, I cannot reproduce it locally - when breaking connection to RabbitMQ, RR always resumes pipeline. What is strange that, when this case occurs, last message of log claims, that redialer exited. Is this expected behaviour? Is there something I can do, to automatically resume pipeline, once RR restores connection? Is there a difference if pipelines are declared dynamically (as currently) or statically, in .rr.yaml file?

```
{"level":"debug","ts":1727514286185636176,"logger":"amqp","msg":"delivery channel was closed, leaving the AMQP listener"}
{"level":"debug","ts":1727514286185643119,"logger":"amqp","msg":"number of listeners","listeners":0}
{"level":"error","ts":1727514286185707689,"logger":"amqp","msg":"consume channel close","error":"Exception (504) Reason: \"channel/connection is not open\""}
{"level":"error","ts":1727514286185767441,"logger":"amqp","msg":"amqp connection closed","error":"Exception (504) Reason: \"channel/connection is not open\""}
{"level":"error","ts":1727514286185779093,"logger":"amqp","msg":"pipeline connection was closed, redialing","error":"Exception (501) Reason: \"read tcp <redacted>:38128-><redacted>:5672: read: connection reset by peer\"","pipeline":"action","driver":"amqp","start":1727514286185774725}
{"level":"info","ts":1727514286228254674,"logger":"amqp","msg":"amqp dial was succeed. trying to redeclare queues and subscribers"}
{"level":"info","ts":1727514286280420208,"logger":"amqp","msg":"queues and subscribers was redeclared successfully"}
{"level":"info","ts":1727514286280465312,"logger":"amqp","msg":"connection was successfully restored","pipeline":"action","driver":"amqp","start":1727514286185774725,"elapsed":94}
{"level":"info","ts":1727514286280484528,"logger":"amqp","msg":"redialer restarted"}
{"level":"debug","ts":1727514286280503032,"logger":"amqp","msg":"exited from redialer"}

$ ./rr workers
Workers of [jobs]:
+---------+-----------+---------+---------+---------+--------------------+
|   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
+---------+-----------+---------+---------+---------+--------------------+
| 2719733 | ready     |      23 | 42 MB   |    0.00 | 8 hours ago        |
| 2719734 | ready     |      22 | 42 MB   |    0.00 | 8 hours ago        |
| 2719735 | ready     |      22 | 43 MB   |    0.00 | 8 hours ago        |
| 2719736 | ready     |      23 | 44 MB   |    0.00 | 8 hours ago        |
| 2719737 | ready     |      23 | 40 MB   |    0.00 | 8 hours ago        |
| 2719738 | ready     |      23 | 42 MB   |    0.00 | 8 hours ago        |
| 2719739 | ready     |      23 | 42 MB   |    0.00 | 8 hours ago        |
| 2719740 | ready     |      23 | 43 MB   |    0.00 | 8 hours ago        |
+---------+-----------+---------+---------+---------+--------------------+
Jobs of [jobs]:
+----------------+----------------+--------+---------------------+--------+---------+----------+
|     STATUS     |    PIPELINE    | DRIVER |        QUEUE        | ACTIVE | DELAYED | RESERVED |
+----------------+----------------+--------+---------------------+--------+---------+----------+
| PAUSED/STOPPED | action         | amqp   | 1306.action         | 0      | 0       | 0        |
| PAUSED/STOPPED | callback       | amqp   | callback            | 0      | 0       | 0        |
| READY          | flow           | amqp   | 1306.flow           | 0      | 0       | 0        |
+----------------+----------------+--------+---------------------+--------+---------+----------+
```

.rr.yaml:
```yaml
version: '3'
rpc:
  listen: 'tcp://127.0.0.1:6001'
jobs:
  timeout: 172800
  pool:
    num_workers: 1
  consume:
    - flow
    - action
server:
  command: 'php app.php'
  relay: pipes
  env:
    - XDEBUG_SESSION: '1'
metrics:
  address: '127.0.0.1:2112'
amqp:
  addr: 'amqp://<redacted>:<redacted>@<redacted>:5672/'
endure:
  log_level: error
  watchdog_sec: 60
```

**Comment by rustatian** [comment]
Hey @algirdasci 👋 
Interesting case... Could you please post here a RR version as well?

**Comment by algirdasci** [comment]
Sure! I'm using latest RR build:
```
$ ./rr --version
rr version 2024.2.1 (build time: 2024-09-12T16:25:04+0000, go1.23.1), OS: linux, arch: amd64
```

**Reply by rustatian** [reply]
Thanks 👍 
`{"level":"debug","ts":1727514286280503032,"logger":"amqp","msg":"exited from redialer"}` this is ok, but I think I need to update these logs to precisely specify what part of redialer have exited.

But, I suspect, that there is should be an additional debug log about some problem. Could you please turn on debug logging?
Also, you may use `resume` method from the PHP to directly resume the pipeline.

**Reply by algirdasci** [reply]
I've created scheduled job just to resume paused pipelines, thanks for advice. 

What debug logging should I enable? I think I already enabled debug logging pretty much everywhere.

**Reply by rustatian** [reply]
Oh, yeah, I missed that you have `endure.log_level`. But, you don't have a logger in the configuration (rr.yaml). Keep in mind, that enabling debug logging will cause some performance penalty (~5-10%).

I double-checked the code and I think that I see where the issue might happen. Scheduled job should help you, but I'll add (for 2024.3) a more specific logs to detect the failure.

**Reply by algirdasci** [reply]
Sorry, must have cut out the `logs` part while copying, this is my logging config in .rr.yaml:
```yaml
...
logs:
  mode: production
  file_logger_options:
    log_output: "node.log"
    max_size: 10
    max_age: 24
    max_backups: 10
    compress: true
...
```

I'll monitor situation with scheduled job resume and let you know if I'll notice some issues.

**Reply by rustatian** [reply]
Sure, thanks 👍 Ping me in this thread if you find a some tricky error or just have a question about some behavior 😃

**Reply by algirdasci** [reply]
Thank you, @rustatian 

Actually, I do have a question. It is related with poor/unstable connection to RabbitMQ. Is it possible to implement [Publisher Confirm](https://www.rabbitmq.com/docs/confirms#publisher-confirms) logic to AMQP driver? Like here in this [example](https://github.com/rabbitmq/amqp091-go/blob/main/_examples/producer/producer.go#L116). Or maybe it is possible via gRPC to send/receive confirmations?

The problem is, that the app I'm currently developing and running on RR is part of the bigger action flow. The master app initializes flow by sending message to RR. Then, RR consumes this message and starts long running task. Then RR periodically checks task progress and reports back to master app. The check is performed by consuming message and if the long running task is still running, such message is pushed again to queue. The problem is that with unstable connection such messages in some cases gets lost and the flow becomes stuck. 
The confirmation, that message was pushed to queue was successful would greatly improve message delivery stability.

**Reply by rustatian** [reply]
Sure, so, the idea is to turn on the confirmation mode via configuration (pipeline) and confirm every message and if not confirmed, reject such a message with an error. This is actually a good proposal, could you please create a Feature Request for that?

**Reply by algirdasci** [reply]
@rustatian I got another interesting case. Today we had couple of RabbitMQ node crashes, after which the pipeline seems to be resumed, but trying to publish a message to the queue fails with error:

```
{"level":"error","ts":1727781359895138763,"logger":"jobs","msg":"job push error","ID":"4192f827-2acb-4df9-a629-d3541240f409","pipeline":"flow","driver":"amqp","start":1727781359895044918,"elapsed":0,"error":"amqp_driver_push:\n\tamqp_driver_handle_item: Exception (504) Reason: \"channel/connection is not open\""}
```
```
$ ./rr workers
Workers of [jobs]:
+---------+-----------+---------+---------+---------+--------------------+
|   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
+---------+-----------+---------+---------+---------+--------------------+
|  622902 | ready     |      21 | 53 MB   |    0.01 | 3 hours ago        |
|  622903 | ready     |      22 | 53 MB   |    0.01 | 3 hours ago        |
|  622904 | ready     |      22 | 51 MB   |    0.00 | 3 hours ago        |
|  622905 | ready     |      22 | 52 MB   |    0.00 | 3 hours ago        |
|  622906 | ready     |      22 | 51 MB   |    0.01 | 3 hours ago        |
|  622907 | ready     |      22 | 51 MB   |    0.00 | 3 hours ago        |
|  622908 | ready     |      21 | 51 MB   |    0.01 | 3 hours ago        |
|  622909 | ready     |      22 | 51 MB   |    0.00 | 3 hours ago        |
+---------+-----------+---------+---------+---------+--------------------+
Jobs of [jobs]:
+----------------+----------------+--------+---------------------+--------+---------+----------+
|     STATUS     |    PIPELINE    | DRIVER |        QUEUE        | ACTIVE | DELAYED | RESERVED |
+----------------+----------------+--------+---------------------+--------+---------+----------+
| READY          | action         | amqp   | 1169.action         | 0      | 0       | 0        |
| PAUSED/STOPPED | callback       | amqp   | callback            | 0      | 0       | 0        |
| READY          | flow           | amqp   | 1169.flow           | 0      | -1      | 0        |
+----------------+----------------+--------+---------------------+--------+---------+----------+
```

**Reply by rustatian** [reply]
Oh yeah, and you see this `-1` in the DELAYED?

**Reply by rustatian** [reply]
Resume operation might fail if RR AMQP plugin currently is trying to redial. If you have logs, it would be really nice to see them 😃

**Reply by algirdasci** [reply]
Yes. It happens from time to time.

Just to give more context, the `flow` pipeline consumes messages to check long running task progress. If the progress is not finished then the consumed job pushes new delayed message with same payload. So maybe when the delayed message pushing fails, then the counter is set to -1?

**Reply by rustatian** [reply]
Yeah, the number in the delayed column is not from the rabbitmq broker (they don't provide such an info). It is just our approximate counter.

**Reply by algirdasci** [reply]
I'm attaching the log. The resume operation did not run, because RR show the pipeline as resumed. I mean the `flow` pipeline, the `callback` pipeline is for publishing only and RR does not consume its messages.

[crash.log](https://github.com/user-attachments/files/17210316/crash.log)

**Reply by rustatian** [reply]
Ok, I checked your logs, it looks like, that amqp driver (rabbitmq amqp driver) did not send a disconnection failure to start a reconnection routine. I'll double check this case.

**Reply by algirdasci** [reply]
Hey, @rustatian how are you doing? While I managed to overcome the message publishing issues, there is one issue that does not let me sleep at night. Sometimes, after connection disruptions, RoadRunner Jobs does not restore connection properly and I cannot resume pipeline:
```
2024-10-10T09:38:56.852285+00:00] roadrunner.ERROR: Spiral\RoadRunner\Jobs\Exception\JobsException: Error 'rpc_stats:  jobs_plugin_drivers_state:      amqp_driver_state: Exception (504) Reason: "channel/connection is not open"' on tcp://127.0.0.1:6001 in /opt/app/releases/600ad9e2d8ea26dbb16f713ccb057ea722b07272/vendor/spiral/roadrunner-jobs/src/Queue.php at line 162 [] []
[Spiral\RoadRunner\Jobs\Exception\JobsException]
Error 'rpc_stats:
        jobs_plugin_drivers_state:
        amqp_driver_state: Exception (504) Reason: "channel/connection is not open"' on tcp://127.0.0.1:6001
in vendor/spiral/roadrunner-jobs/src/Queue.php:162

Previous: [Spiral\Goridge\RPC\Exception\ServiceException]
Error 'rpc_stats:
        jobs_plugin_drivers_state:
        amqp_driver_state: Exception (504) Reason: "channel/connection is not open"' on tcp://127.0.0.1:6001
in vendor/spiral/goridge/src/RPC/AbstractRPC.php:72
```
trying resume via `./rr jobs --resume flow` the command just hangs. Any advice how to restore connection?

**Reply by rustatian** [reply]
Hey @algirdasci 👋 
Hm, it looks like some deadlock in the retry logic. I mean, if you use a huge timeout, then, backoff strategy might choose long periods between retries. And the resume lock would be locked, which would lead to `./rr jobs --resume` command to hang.

**Reply by algirdasci** [reply]
I use high `redial_timeout` value, to be sure, that message consuming would resume eventually. What would happen if I use lower `redial_timeout`, like 30 seconds, but the connection will not be restored during that time?

**Reply by rustatian** [reply]
In that case, in the latest versions, RR will stop the pipeline and try to recreate it from scratch. And you'll see the following error message in the logs: `"backoff operation failed, pipeline will be recreated`. I think, for you, would be safer to use 2-3 minutes timeout and let RR recreate a pipeline on fail.

**Reply by algirdasci** [reply]
Thanks for the advice, I will test that scenario.

**Reply by algirdasci** [reply]
When I set default timeout values, the pipeline does not get recreated. RR just says `pipeline will be recreated` but nothing happens. How to reproduce:

1. Start RoadRunner
2. Stop or pause RabbitMQ container
3. Wait until timeout is reached
4. Start or resume RabbitMQ container


```bash
2024-10-10T14:03:14+0000        DEBUG   rpc             plugin was started      {"address": "tcp://127.0.0.1:6001", "list of the plugins with RPC methods:": ["lock", "kv", "metrics", "resetter", "informer", "app", "jobs"]}
2024-10-10T14:03:14+0000        DEBUG   amqp            pipeline was started    {"driver": "amqp", "pipeline": "rpc", "start": "2024-10-10T14:03:14+0000", "elapsed": 4}
2024-10-10T14:03:15+0000        DEBUG   server          worker is allocated     {"pid": 84205, "max_execs": 0, "internal_event_name": "EventWorkerConstruct"}
2024-10-10T14:03:44+0000        DEBUG   amqp            delivery channel was closed, leaving the AMQP listener
2024-10-10T14:03:44+0000        DEBUG   amqp            exited from redialer
2024-10-10T14:03:44+0000        DEBUG   amqp            number of listeners     {"listeners": 0}
2024-10-10T14:03:44+0000        ERROR   amqp            amqp connection closed  {"error": "Exception (504) Reason: \"channel/connection is not open\""}
2024-10-10T14:03:44+0000        ERROR   amqp            pipeline connection was closed, redialing       {"error": "Exception (501) Reason: \"read tcp 172.23.0.2:57384->192.168.65.254:5672: i/o timeout\"", "pipeline": "rpc", "driver": "amqp", "start": "2024-10-10T14:03:44+0000"}
2024-10-10T14:04:45+0000        ERROR   amqp            backoff operation failed, pipeline will be recreated    {"error": "amqp_driver_redial: Exception (501) Reason: \"read tcp 172.23.0.2:44474->192.168.65.254:5672: i/o timeout\""}

Meanwhile from another terminal:

root@7756dcc75962:/app# ./rr workers
Workers of [jobs]:
+---------+-----------+---------+---------+---------+--------------------+
|   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
+---------+-----------+---------+---------+---------+--------------------+
|   84205 | ready     |       0 | 44 MB   |    0.05 | 14 minutes ago     |
+---------+-----------+---------+---------+---------+--------------------+

root@7756dcc75962:/app# ./rr jobs --resume rpc
jobs_plugin_rmq_queue_declare: Exception (504) Reason: "channel/connection is not open"
```

For testing I'm using following .rr.yaml. Keep in mind, that `rpc` pipeline is declared dynamically:
```yaml
version: '3'
rpc:
  listen: 'tcp://127.0.0.1:6001'
kv:
  local:
    driver: memory
    config:
      interval: 60
jobs:
  pool:
    num_workers: 1
  consume:
    - rpc
server:
  command: 'php app.php'
  relay: pipes
  env:
    - XDEBUG_SESSION: '1'
metrics:
  address: '127.0.0.1:2112'
amqp:
  addr: 'amqp://guest:guest@host.docker.internal:5672/'
endure:
  log_level: error
  watchdog_sec: 60
```

**Reply by rustatian** [reply]
I'll double-check that, need to write an additional test for that case.

**Reply by rustatian** [reply]
I did not forget about this case @algirdasci, just need some more time to find the problem (friendly note 😃 )

**Reply by algirdasci** [reply]
all good, no rush :) thanks for your work!

**Reply by rustatian** [reply, marked as answer]
I think I found the reason. There is might be a case, when the channel for messages delivery could be closed by the RabbitMQ server, without connection interruption. In that case, RR will close the delivery channel but does not create a new one, since the trigger for redial is broken connection.
I wrote a new Redialer which can be applied to channels + connections and recreate a delivery channel in that case. Expected release: 2024.3.1. Could you please create a bug ticket for that?

**Reply by algirdasci** [reply]
Great find, @rustatian! I've created [bug ticket](https://github.com/roadrunner-server/roadrunner/issues/2091) for it.

**Reply by rustatian** [reply]
I still remember about your case @algirdasci 😃 
I've added more logs and tune the closing logic a bit. Will be released soon.
